open_spiel/games/chess/chess_board.cc

```diff
@@ -1440,9 +1430,93 @@ std::string ChessBoard<kBoardSize>::ToFEN() const {
   return fen;
 }
 
-template <uint32_t kBoardSize>
-void ChessBoard<kBoardSize>::set_square(Square sq, Piece piece) {
-  static const ZobristTableU64<kBoardSize * kBoardSize, 3, 7> kZobristValues(
+std::string ChessBoard::ToDarkFEN(const std::array<bool, kMaxBoardSize * kMaxBoardSize>& observability_table,
+                                  Color color) const {
+  std::string fen;
+
+  // 1. encode the board.
+  for (int8_t rank = board_size_ - 1; rank >= 0; --rank) {
+    int num_empty = 0;
+    for (int8_t file = 0; file < board_size_; ++file) {
+
+      size_t index = SquareToIndex_(chess::Square{file, rank});
+      if (!observability_table[index]) {
+        if (num_empty > 0) {
+          fen += std::to_string(num_empty);
+          num_empty = 0;
+        }
+        fen += '?';
+      } else {
+        auto piece = at(chess::Square{file, rank});
+        if (piece == chess::kEmptyPiece) {
+          ++num_empty;
+        } else {
+          if (num_empty > 0) {
+            fen += std::to_string(num_empty);
+            num_empty = 0;
+          }
+          fen += piece.ToString();
+        }
+      }
+    }
+    if (num_empty > 0) {
+      fen += std::to_string(num_empty);
+    }
+    if (rank > 0) {
+      fen += "/";
+    }
+  }
+
+  // 2. color to play.
+  fen += " " + (ToPlay() == chess::Color::kWhite ? std::string("w") : std::string("b"));
+
+  // 3. by castling rights.
+  fen += " ";
+  std::string castling_rights;
+  if (color == chess::Color::kWhite) {
+    if (CastlingRight(chess::Color::kWhite, chess::CastlingDirection::kRight)) {
+      castling_rights += "K";
+    }
+    if (CastlingRight(chess::Color::kWhite, chess::CastlingDirection::kLeft)) {
+      castling_rights += "Q";
+    }
+  } else {
+    if (CastlingRight(chess::Color::kBlack, chess::CastlingDirection::kRight)) {
+      castling_rights += "k";
+    }
+    if (CastlingRight(chess::Color::kBlack, chess::CastlingDirection::kLeft)) {
+      castling_rights += "q";
+    }
+  }
+  fen += castling_rights.empty() ? std::string("-") : castling_rights;
+
+  // 4. en passant square
```

Comment:
How does en-passant work with limited observations? The player can "suddenly" lose a piece right? Is it encoded in observations?